src/System.Management.Automation/engine/runtime/Operations/ClassOps.cs

```diff
@@ -42,26 +42,19 @@ internal SessionStateKeeper()
 
         internal void RegisterRunspace()
         {
-            SessionStateInternal ssInMap = null;
-            Runspace rsToUse = Runspace.DefaultRunspace;
-            SessionStateInternal ssToUse = rsToUse.ExecutionContext.EngineSessionState;
+            Runspace runspaceToUse = Runspace.DefaultRunspace;
+            SessionStateInternal sessionStateToUse = runspaceToUse.ExecutionContext.EngineSessionState;
 
             // Different threads will operate on different key/value pairs (default-runspace/session-state pairs),
             // and a ConditionalWeakTable itself is thread safe, so there won't be race condition here.
-            if (!_stateMap.TryGetValue(rsToUse, out ssInMap))
-            {
-                // If the key doesn't exist yet, add it
-                _stateMap.Add(rsToUse, ssToUse);
-            }
-            else if (!ssInMap.Equals(ssToUse))
-            {
-                // If the key exists but the corresponding value is not what we should use, then remove the key/value pair and add the new pair.
-                // This could happen when a powershell class is defined in a module and the module gets reloaded. In such case, the same TypeDefinitionAst
-                // instance will get reused, but should be associated with the SessionState from the new module, instead of the one from the old module.
-                _stateMap.Remove(rsToUse);
-                _stateMap.Add(rsToUse, ssToUse);
-            }
+
+            // If the key exists but the corresponding value is not what we should use, then update the key/value pair.
+            // This could happen when a powershell class is defined in a module and the module gets reloaded.
+            // In such case, the same TypeDefinitionAst instance will get reused,
+            // but should be associated with the SessionState from the new module, instead of the one from the old module.
             // If the key exists and the corresponding value is the one we should use, then do nothing.
+
+            _stateMap.AddOrUpdate(runspaceToUse, sessionStateToUse);
```

Comment:
This change is not right. Update is not need for most of the cases, but `AddOrUpdate` will always update the value, even though the existing value is already what we want.

Reply:
> for most of the cases

My mistake is that I skipped that AddOrUpdate does a lock. 
Update itself is not expensive for references.
I think we could only replace Remove()+ Add() (lines 61-62 in original code) with single lock call AddOrUpdate() - this is twice faster (both Remove() and Add() do lock and FindEntry) and exclude a race condition.

Reply:
> replace Remove()+ Add() (lines 61-62 in original code) with single lock call AddOrUpdate()

This sounds about right. Please go ahead making the change.
Also, maybe change `!ssInMap.Equals(ssToUse)` to `ssInMap != ssToUse`

Reply:
Done.